core/src/main/java/org/elasticsearch/index/mapper/TypeFieldMapper.java

```diff
@@ -204,6 +208,66 @@ public String toString(String field) {
 
     }
 
+    /**
+     * Specialization for a disjunction over many _type
+     */
+    public static class TypesQuery extends Query {
+        // Same threshold as TermsQuery
+        private static final int BOOLEAN_REWRITE_TERM_COUNT_THRESHOLD = 16;
+
+        private final BytesRef[] types;
+
+        public TypesQuery(BytesRef... types) {
+            this.types = Objects.requireNonNull(types);
+        }
+
+        @Override
+        public Query rewrite(IndexReader reader) throws IOException {
+            final int threshold = Math.min(BOOLEAN_REWRITE_TERM_COUNT_THRESHOLD, BooleanQuery.getMaxClauseCount());
+            if (types.length <= threshold) {
+                BooleanQuery.Builder bq = new BooleanQuery.Builder();
+                for (BytesRef type : types) {
+                    Query tq = new TypeQuery(BytesRef.deepCopyOf(type)).rewrite(reader);
+                    if (tq instanceof MatchAllDocsQuery) {
+                        // The reader contains a single _type
+                        return tq;
+                    }
+                    if (tq instanceof ConstantScoreQuery) {
+                        bq.add(((ConstantScoreQuery) tq).getQuery(), BooleanClause.Occur.SHOULD);
+                    }
```

Comment:
Actually maybe we can do better in a TypesQuery? For instance imagine there are two types `foo` and `bar`. If both types are in the list then we could rewrite as a match_all query?

I guess we could either check that the types here are a superset of the types returned by MapperService.types() or compute the sum of the doc freqs of each term and rewrite to a match_all query if the sum is equal to `maxDoc`?

Then if TypesQuery is able to do this, we do not need TypeQuery anymore and can use TypesQuery all the time?